test/HttpCacheListenerTest.php

```diff
@@ -665,4 +725,73 @@ public function testSetVary(array $cacheConfig, array $headers, array $exHeaders
 
         $this->assertSame($exHeaders, $headers->toArray());
     }
+
+
+    /**
+     * @covers \ZF\HttpCache\HttpCacheListener::setvary
+     * @dataProvider setEtagDataProvider
+     *
+     * @param array $cacheConfig
+     * @param array $headers
+     * @param array $exHeaders
+     */
+    public function testSetETag(array $cacheConfig, array $headers, array $exHeaders)
+    {
+        $this->instance->setCacheConfig($cacheConfig);
+
+        $response = new HttpResponse();
+        $headers  = $response->getHeaders()
+            ->addHeaders($headers);
+
+        $this->instance->setEtag(new HttpRequest(), $response);
+
+        $this->assertSame($exHeaders, $headers->toArray());
+    }
+
+    public function testSetETagGenerator()
```

Comment:
Also write a test that demonstrates that the ETag generation is skipped if the appropriate conditions are not met.